riverdriver/riverpgxv5/river_pgx_v5_driver.go

```diff
@@ -99,34 +447,133 @@ func (t *ExecutorTx) Rollback(ctx context.Context) error {
 	return t.tx.Rollback(ctx)
 }
 
-func interpretError(err error) error {
-	if errors.Is(err, pgx.ErrNoRows) {
-		return riverdriver.ErrNoRows
+type Listener struct {
+	conn   *pgxpool.Conn
+	dbPool *pgxpool.Pool
+	mu     sync.RWMutex
+}
+
+func (l *Listener) Close(ctx context.Context) error {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+
+	if l.conn == nil {
+		return nil
+	}
+
+	if err := l.conn.Conn().Close(ctx); err != nil {
+		return err
+	}
+	l.conn.Release()
+	l.conn = nil
+	return nil
+}
+
+func (l *Listener) Connect(ctx context.Context) error {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+
+	if l.conn != nil {
+		return errors.New("connection already established")
+	}
+
+	conn, err := l.dbPool.Acquire(ctx)
+	if err != nil {
+		return err
 	}
+
+	l.conn = conn
+	return nil
+}
+
+func (l *Listener) Listen(ctx context.Context, topic string) error {
+	l.mu.RLock()
+	defer l.mu.RUnlock()
+
+	_, err := l.conn.Exec(ctx, "LISTEN "+topic)
 	return err
 }
 
-func mapMigrations(migrations []*dbsqlc.RiverMigration) []*riverdriver.Migration {
-	if migrations == nil {
-		return nil
+func (l *Listener) Ping(ctx context.Context) error {
+	l.mu.RLock()
+	defer l.mu.RUnlock()
+
+	return l.conn.Ping(ctx)
+}
+
+func (l *Listener) Unlisten(ctx context.Context, topic string) error {
+	l.mu.RLock()
+	defer l.mu.RUnlock()
+
+	_, err := l.conn.Exec(ctx, "UNLISTEN "+topic)
+	return err
+}
+
+func (l *Listener) WaitForNotification(ctx context.Context) (*riverdriver.Notification, error) {
+	l.mu.RLock()
+	defer l.mu.RUnlock()
+
+	notification, err := l.conn.Conn().WaitForNotification(ctx)
+	if err != nil {
+		return nil, err
 	}
 
-	return mapSlice(migrations, func(m *dbsqlc.RiverMigration) *riverdriver.Migration {
-		return &riverdriver.Migration{
-			ID:        int(m.ID),
-			CreatedAt: m.CreatedAt,
-			Version:   int(m.Version),
-		}
-	})
+	return &riverdriver.Notification{
+		Topic:   notification.Channel,
+		Payload: notification.Payload,
+	}, nil
 }
 
-// mapSlice manipulates a slice and transforms it to a slice of another type.
-func mapSlice[T any, R any](collection []T, mapFunc func(T) R) []R {
-	result := make([]R, len(collection))
+func attemptErrorFromInternal(e *dbsqlc.AttemptError) rivertype.AttemptError {
+	return rivertype.AttemptError{
+		At:      e.At,
+		Attempt: int(e.Attempt),
+		Error:   e.Error,
+		Trace:   e.Trace,
+	}
+}
 
-	for i, item := range collection {
-		result[i] = mapFunc(item)
+func interpretError(err error) error {
+	if errors.Is(err, pgx.ErrNoRows) {
+		return rivertype.ErrNotFound
+	}
+	return err
+}
+
+func jobRowFromInternal(internal *dbsqlc.RiverJob) *rivertype.JobRow {
+	return &rivertype.JobRow{
+		ID:          internal.ID,
```

Comment:
May as well fully sort this struct declaration now, same for `migrationFromInternal` below

Reply:
I _think_ it is. I'm sort of using the convention of `ID` at the top, and then alphabetical from there on out.